// polarity-integration-utils/dist/lib/testing/test-factory/test-do-lookup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.testDoLookup = void 0;
const zod_types_1 = require("../../zod-types");
const create_mock_integration_context_1 = require("../enhanced-utils/create-mock-integration-context");
const testDoLookup = (integration, description, { entities, options, expected }) => {
    test(description, async () => {
        const mockLogger = {
            info: jest.fn(),
            debug: jest.fn(),
            trace: jest.fn(),
            warn: jest.fn(),
            error: jest.fn(),
            fatal: jest.fn()
        };
        await integration.startup(mockLogger);
        const context = (0, create_mock_integration_context_1.createMockIntegrationContext)();
        const result = await integration.doLookup(entities, options, context);
        const validationResult = zod_types_1.DoLookupResultSchema.safeParse(result);
        expect(validationResult.success).toBe(true);
        expect(result).toEqual(expected);
    });
};
exports.testDoLookup = testDoLookup;
