// polarity-integration-utils/dist/lib/testing/test-factory/test-on-message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.testOnMessage = void 0;
const create_mock_integration_context_1 = require("../enhanced-utils/create-mock-integration-context");
const testOnMessage = (integration, description, { command, args, expected, options = {} }) => {
    test(description, async () => {
        const mockLogger = {
            info: jest.fn(),
            debug: jest.fn(),
            trace: jest.fn(),
            warn: jest.fn(),
            error: jest.fn(),
            fatal: jest.fn()
        };
        await integration.startup(mockLogger);
        const context = (0, create_mock_integration_context_1.createMockIntegrationContext)();
        if (typeof integration.onMessage !== 'function') {
            throw new Error('Integration does not have an onMessage method.');
        }
        const result = await integration.onMessage({ command, ...args }, options, context);
        expect(result).toEqual(expected);
    });
};
exports.testOnMessage = testOnMessage;
