// polarity-integration-utils/dist/lib/zod-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntegrationConfigSchema = exports.DoLookupResultSchema = exports.ResultSchema = exports.EntitySchema = exports.EntityTypeSchema = exports.StandardEntityTypeSchema = exports.ChannelSchema = exports.ValidateOptionsUserOptionsSchema = exports.ValidateOptionsUserOptionSchema = exports.DoLookupUserOptionsSchema = exports.PossibleUserOptionValueSchema = exports.DropdownUserOptionValueSchema = exports.ValidationErrorSchema = void 0;
const zod_1 = require("zod");
exports.ValidationErrorSchema = zod_1.z.object({
    key: zod_1.z.string(),
    message: zod_1.z.string()
});
exports.DropdownUserOptionValueSchema = zod_1.z.object({
    display: zod_1.z.string(),
    value: zod_1.z.string()
});
exports.PossibleUserOptionValueSchema = zod_1.z.union([
    zod_1.z.undefined(),
    zod_1.z.string(),
    zod_1.z.number(),
    zod_1.z.boolean(),
    exports.DropdownUserOptionValueSchema,
    zod_1.z.array(exports.DropdownUserOptionValueSchema)
]);
exports.DoLookupUserOptionsSchema = zod_1.z.record(zod_1.z.string(), exports.PossibleUserOptionValueSchema);
exports.ValidateOptionsUserOptionSchema = zod_1.z.object({
    integration_id: zod_1.z.string().optional(),
    key: zod_1.z.string(),
    value: exports.PossibleUserOptionValueSchema,
    user_can_edit: zod_1.z.boolean().optional(),
    admin_only: zod_1.z.boolean().optional()
});
exports.ValidateOptionsUserOptionsSchema = zod_1.z.record(zod_1.z.string(), exports.ValidateOptionsUserOptionSchema);
exports.ChannelSchema = zod_1.z.object({
    channel_name: zod_1.z.string(),
    id: zod_1.z.number()
});
exports.StandardEntityTypeSchema = zod_1.z.union([
    zod_1.z.literal('IP'),
    zod_1.z.literal('IPv4'),
    zod_1.z.literal('IPv4CIDR'),
    zod_1.z.literal('IPv6'),
    zod_1.z.literal('MAC'),
    zod_1.z.literal('MD5'),
    zod_1.z.literal('SHA1'),
    zod_1.z.literal('SHA256'),
    zod_1.z.literal('cve'),
    zod_1.z.literal('domain'),
    zod_1.z.literal('email'),
    zod_1.z.literal('hash'),
    zod_1.z.literal('string'),
    zod_1.z.literal('url')
]);
exports.EntityTypeSchema = zod_1.z.union([
    exports.StandardEntityTypeSchema,
    zod_1.z.literal('*'),
    zod_1.z.literal('custom'),
    zod_1.z.custom((val) => /^custom\..+$/.test(val))
]);
exports.EntitySchema = zod_1.z.object({
    value: zod_1.z.string(),
    types: zod_1.z.array(exports.EntityTypeSchema),
    type: exports.EntityTypeSchema,
    requestContext: zod_1.z.object({
        requestType: zod_1.z.literal('onDemand'),
        isUserInitiated: zod_1.z.boolean()
    }),
    longitude: zod_1.z.number(),
    latitude: zod_1.z.number(),
    isURL: zod_1.z.boolean(),
    isSHA512: zod_1.z.boolean(),
    isSHA256: zod_1.z.boolean(),
    isSHA1: zod_1.z.boolean(),
    isPrivateIP: zod_1.z.boolean(),
    isMD5: zod_1.z.boolean(),
    isIPv6: zod_1.z.boolean(),
    isIPv4: zod_1.z.boolean(),
    isIP: zod_1.z.boolean(),
    isHex: zod_1.z.boolean(),
    isHash: zod_1.z.boolean(),
    isHTMLTag: zod_1.z.boolean(),
    isEmail: zod_1.z.boolean(),
    isDomain: zod_1.z.boolean(),
    hashType: zod_1.z.union([
        zod_1.z.literal('md5'),
        zod_1.z.literal('sha1'),
        zod_1.z.literal('sha256'),
        zod_1.z.literal('sha512'),
        zod_1.z.literal('')
    ]),
    displayValue: zod_1.z.string(),
    channels: zod_1.z.array(exports.ChannelSchema),
    IPType: zod_1.z.union([zod_1.z.literal('IPv4'), zod_1.z.literal('IPv6'), zod_1.z.literal('')])
});
exports.ResultSchema = zod_1.z.object({
    entity: exports.EntitySchema,
    displayValue: zod_1.z.string().optional(),
    data: zod_1.z.object({
        summary: zod_1.z.array(zod_1.z.string()),
        details: zod_1.z.unknown()
    })
});
exports.DoLookupResultSchema = zod_1.z.array(exports.ResultSchema);
const SelectOptionItemSchema = zod_1.z.object({
    value: zod_1.z.string(),
    display: zod_1.z.string()
});
// Schema for the 'options' array items (User Configuration)
const OptionSchema = zod_1.z.discriminatedUnion('type', [
    zod_1.z.object({
        type: zod_1.z.enum(['text', 'password']),
        key: zod_1.z.string(),
        name: zod_1.z.string(),
        description: zod_1.z.string().optional(),
        default: zod_1.z.string().nullable(),
        userCanEdit: zod_1.z.boolean().optional(),
        adminOnly: zod_1.z.boolean().optional()
    }),
    zod_1.z.object({
        type: zod_1.z.literal('boolean'),
        key: zod_1.z.string(),
        name: zod_1.z.string(),
        description: zod_1.z.string().optional(),
        default: zod_1.z.boolean().nullable(),
        userCanEdit: zod_1.z.boolean().optional(),
        adminOnly: zod_1.z.boolean().optional()
    }),
    zod_1.z.object({
        type: zod_1.z.literal('number'),
        key: zod_1.z.string(),
        name: zod_1.z.string(),
        description: zod_1.z.string().optional(),
        default: zod_1.z.number().nullable(),
        userCanEdit: zod_1.z.boolean().optional(),
        adminOnly: zod_1.z.boolean().optional()
    }),
    zod_1.z.object({
        type: zod_1.z.literal('select'),
        key: zod_1.z.string(),
        name: zod_1.z.string(),
        description: zod_1.z.string().optional(),
        // Default can be a single object or an array of objects depending on 'multiple'
        default: zod_1.z
            .union([
            SelectOptionItemSchema,
            zod_1.z.array(SelectOptionItemSchema),
            zod_1.z.string() // Occasionally a simple string value is used for default
        ])
            .nullable(),
        options: zod_1.z.array(SelectOptionItemSchema),
        multiple: zod_1.z.boolean().optional(),
        userCanEdit: zod_1.z.boolean(),
        adminOnly: zod_1.z.boolean()
    })
]);
// Schema for custom data types defined within the config
const CustomTypeSchema = zod_1.z.object({
    type: zod_1.z.literal('custom').optional(), // Sometimes present in dataTypes
    name: zod_1.z.string().optional(),
    description: zod_1.z.string().optional(),
    key: zod_1.z.string(),
    regex: zod_1.z.string(),
    editable: zod_1.z.boolean().optional(),
    enabled: zod_1.z.boolean().optional()
});
// Component definition for block/summary
const ComponentPathSchema = zod_1.z.object({
    file: zod_1.z.string()
});
const ViewComponentSchema = zod_1.z.object({
    component: ComponentPathSchema,
    template: ComponentPathSchema
});
exports.IntegrationConfigSchema = zod_1.z.object({
    polarityIntegrationUuid: zod_1.z.uuid(),
    name: zod_1.z.string(),
    acronym: zod_1.z.string().max(10),
    description: zod_1.z.string().optional(),
    defaultColor: zod_1.z.string().optional(),
    // Entity and Data types
    entityTypes: zod_1.z.array(zod_1.z.string()).optional(),
    dataTypes: zod_1.z.array(zod_1.z.union([zod_1.z.string(), CustomTypeSchema])).optional(),
    customTypes: zod_1.z.array(CustomTypeSchema).optional(),
    supportsAdditionalCustomTypes: zod_1.z.boolean().optional(),
    // UI/View Configuration
    styles: zod_1.z.array(zod_1.z.string()).optional(),
    block: ViewComponentSchema,
    summary: ViewComponentSchema.optional(),
    // Behavior Configuration
    onDemandOnly: zod_1.z.boolean().optional(),
    copyOnDemand: zod_1.z.boolean().optional(),
    // Logging and Networking
    logging: zod_1.z
        .object({
        level: zod_1.z.enum(['trace', 'debug', 'info', 'warn', 'error', 'fatal'])
    })
        .optional(),
    request: zod_1.z
        .object({
        cert: zod_1.z.string().optional(),
        key: zod_1.z.string().optional(),
        passphrase: zod_1.z.string().optional(),
        ca: zod_1.z.string().optional(),
        proxy: zod_1.z.string().optional(),
        rejectUnauthorized: zod_1.z.boolean().optional()
    })
        .optional(),
    // User Options
    options: zod_1.z.array(OptionSchema).optional()
});
