// polarity-integration-utils/dist/lib/testing/enhanced-utils/create-entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEntity = void 0;
const createEntity = (type, value) => {
    const isDomain = /^[a-zA-Z0-9][a-zA-Z0-9-]{1,61}[a-zA-Z0-9](?:\.[a-zA-Z]{2,})+$/.test(value);
    const isIPv4 = /^(?:[0-9]{1,3}\.){3}[0-9]{1,3}$/.test(value);
    return {
        value,
        displayValue: value,
        type,
        types: [type],
        isDomain,
        isIPv4,
        isIP: isIPv4,
        isEmail: false,
        isHash: false,
        isHex: false,
        isHTMLTag: false,
        isIPv6: false,
        isMD5: false,
        isPrivateIP: false,
        isSHA1: false,
        isSHA256: false,
        isSHA512: false,
        isURL: false,
        latitude: 0,
        longitude: 0,
        hashType: '',
        IPType: isIPv4 ? 'IPv4' : '',
        channels: [],
        requestContext: {
            requestType: 'onDemand',
            isUserInitiated: true
        }
    };
};
exports.createEntity = createEntity;
