// polarity-integration-utils/dist/es/lib/errors/integration-error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntegrationError = void 0;
const parse_error_to_readable_json_1 = require("./parse-error-to-readable-json");
const sanitize_request_options_1 = require("../requests/sanitize-request-options");
/**
 * @public
 */
class IntegrationError extends Error {
    /**
     *  a short, human-readable summary of the problem that SHOULD NOT change from occurrence to occurrence of
     *  the problem except for purposes of localization.
     */
    title;
    /**
     * a human-readable explanation specific to this occurrence of the problem. Like title, this field’s value can be
     * localized
     */
    detail;
    /**
     * a meta object containing non-standard meta-information about the error.
     */
    meta;
    /**
     * The HTTP status code applicable to this error, expressed as a string value.
     */
    status;
    /**
     *  an application-specific error code, expressed as a string value.
     */
    code;
    /**
     * The `cause` property is used to specify the `cause` of the error.  Typically,
     * this property is used to pass through a related Error instance.
     */
    cause;
    /**
     * Additional details related to the error that may help the user troubleshoot the issue.  If set by the user
     * via the Error constructor, the user provided value will override any automated help message set by the
     * Error class.
     */
    help;
    /**
     * Relevant for integration errors involving a network call, the `requestOptions` property
     * details the request options that resulted in the specified error.  The `requestOptions` property will automatically
     * have sensitive authentication headers stripped.
     */
    requestOptions;
    /**
     * Construct a new IntegrationError object.
     * @param message - A string description of the error which is used as the `detail` property on the
     * serialized error.
     * @param properties - Optional properties for the error.
     */
    constructor(message, properties = {}) {
        super(message);
        this.title = properties.title || this.constructor.name;
        this.detail = message;
        this.name = this.constructor.name;
        if (typeof properties.code !== 'undefined') {
            this.code = properties.code;
        }
        if (typeof properties.status !== 'undefined') {
            this.status = properties.status;
        }
        if (properties.cause && properties.cause instanceof Error) {
            this.cause = properties.cause;
        }
        if (typeof properties.requestOptions !== 'undefined') {
            this.requestOptions = (0, sanitize_request_options_1.sanitizeRequestOptions)(properties.requestOptions, properties.requestOptionsToSanitize ? properties.requestOptionsToSanitize : []);
        }
        if (typeof properties.help !== 'undefined') {
            this.help = properties.help;
        }
        if (typeof properties.meta !== 'undefined') {
            this.meta = {
                ...properties.meta
            };
        }
    }
    /**
     * Serializes the error's properties into a POJO.  The order of the
     * properties is preserved when serialized.
     * @returns JSON representation of the error
     */
    toJSON() {
        const props = {
            name: this.name,
            detail: this.detail,
            title: this.title
        };
        if (this.stack) {
            props.stack = this.stack;
        }
        if (this.code) {
            props.code = this.code;
        }
        if (this.status) {
            props.status = this.status;
        }
        if (this.help) {
            props.help = this.help;
        }
        if (this.cause) {
            props.cause = (0, parse_error_to_readable_json_1.parseErrorToReadableJson)(this.cause);
        }
        if (this.requestOptions) {
            props.requestOptions = this.requestOptions;
        }
        if (this.meta && Object.keys(this.meta).length > 0) {
            props.meta = this.meta;
        }
        return props;
    }
}
exports.IntegrationError = IntegrationError;
