// polarity-integration-utils/dist/es/lib/testing/enhanced-utils/validate-integration.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateIntegration = void 0;
const zod_types_1 = require("../../zod-types");
const create_mock_integration_context_1 = require("./create-mock-integration-context");
const zod_1 = __importDefault(require("zod"));
const validateIntegration = (integration, config) => {
    const result = {
        isValid: true,
        errors: [],
        warnings: []
    };
    // Check required functions
    checkStartupFunction(integration, result);
    checkDoLookupFunction(integration, result);
    checkOnMessageFunction(integration, result);
    // Check function signatures
    validateFunctionSignatures(integration, result);
    // Check exported integration object structure (e.g. if it's wrapped in an 'integration' property)
    checkIntegrationObject(integration, result);
    if (config) {
        const configResult = zod_types_1.IntegrationConfigSchema.safeParse(config);
        if (!configResult.success) {
            result.errors.push(`Invalid config.json: ${JSON.stringify(zod_1.default.treeifyError(configResult.error), null, 2)}`);
        }
    }
    result.isValid = result.errors.length === 0;
    if (!result.isValid) {
        throw new Error(`Integration validation failed:\n${result.errors.map((e) => `• ${e}`).join('\n')}`);
    }
    return result;
};
exports.validateIntegration = validateIntegration;
function checkStartupFunction(integration, result) {
    if (typeof integration.startup !== 'function') {
        result.errors.push('Missing required startup function');
    }
    else {
        if (integration.startup.length !== 1) {
            result.warnings.push(`startup function should accept exactly 1 parameter (logger), found ${integration.startup.length}`);
        }
    }
}
function checkDoLookupFunction(integration, result) {
    if (typeof integration.doLookup !== 'function') {
        result.errors.push('Missing required doLookup function');
    }
    else {
        if (integration.doLookup.length !== 3) {
            result.warnings.push(`doLookup function should accept exactly 3 parameters (entities, options, context), found ${integration.doLookup.length}`);
        }
    }
}
function checkOnMessageFunction(integration, result) {
    if (integration.onMessage !== undefined) {
        if (typeof integration.onMessage !== 'function') {
            result.errors.push('onMessage must be a function if present');
        }
        else if (integration.onMessage.length !== 3) {
            result.warnings.push(`onMessage function should accept exactly 3 parameters (payload, options, context), found ${integration.onMessage.length}`);
        }
    }
}
function validateFunctionSignatures(integration, result) {
    const mockLogger = {
        info: () => { },
        debug: () => { },
        warn: () => { },
        error: () => { },
        trace: () => { },
        fatal: () => { }
    };
    const mockContext = (0, create_mock_integration_context_1.createMockIntegrationContext)();
    // Test startup return type
    if (typeof integration.startup === 'function') {
        try {
            const startupResult = integration.startup(mockLogger);
            if (!startupResult || typeof startupResult.then !== 'function') {
                result.errors.push('startup function must return a Promise');
            }
            else {
                // Prevent unhandled rejection if startup throws asynchronously
                startupResult.catch(() => { });
            }
        }
        catch (error) {
            result.warnings.push(`startup function threw error during validation: ${error}`);
        }
    }
    // Test doLookup return type
    if (typeof integration.doLookup === 'function') {
        try {
            const doLookupResult = integration.doLookup([], { validatePayload: false }, mockContext);
            const isValidReturn = Array.isArray(doLookupResult) ||
                doLookupResult === null ||
                doLookupResult === undefined ||
                (doLookupResult &&
                    typeof doLookupResult === 'object' &&
                    'name' in doLookupResult &&
                    doLookupResult.name === 'IntegrationError') ||
                // Check for promise
                (doLookupResult &&
                    typeof doLookupResult.then === 'function');
            if (!isValidReturn) {
                // It might be a promise that resolves to the valid return types, so checking if it's a promise is important.
                // The original code checked constructor name IntegrationError.
                // Let's refine this. doLookup MUST return a Promise.
                if (doLookupResult &&
                    typeof doLookupResult.then === 'function') {
                    // It's a promise, which is good.
                    doLookupResult.catch(() => { });
                }
                else {
                    result.warnings.push(`doLookup should return a Promise, returned: ${typeof doLookupResult}`);
                }
            }
            else if (doLookupResult &&
                typeof doLookupResult.then === 'function') {
                doLookupResult.catch(() => { });
            }
        }
        catch (error) {
            result.warnings.push(`doLookup function threw error during validation: ${error}`);
        }
    }
    // Test onMessage return type
    if (typeof integration.onMessage === 'function') {
        try {
            const onMessageResult = integration.onMessage({}, { validatePayload: false }, mockContext);
            if (!onMessageResult || typeof onMessageResult.then !== 'function') {
                result.errors.push('onMessage function must return a Promise');
            }
            else {
                // Prevent unhandled rejection if onMessage throws asynchronously
                onMessageResult.catch(() => { });
            }
        }
        catch (error) {
            result.warnings.push(`onMessage function threw error during validation: ${error}`);
        }
    }
}
function checkIntegrationObject(integration, result) {
    // Check if someone accidentally exported { integration: { ... } } instead of { ... }
    if ('integration' in integration) {
        const integrationObj = integration.integration;
        if (integrationObj &&
            typeof integrationObj === 'object' &&
            'startup' in integrationObj &&
            typeof integrationObj.startup === 'function' &&
            'doLookup' in integrationObj &&
            typeof integrationObj.doLookup === 'function') {
            result.warnings.push('Detected potential incorrect export structure. The integration seems to be nested under an "integration" property. Ensure you are exporting the integration object directly (e.g. export = { ... }).');
        }
    }
}
